contracts/contracts/Umbra.sol

```diff
@@ -69,72 +133,99 @@ contract Umbra is Ownable {
 
+  /**
+   * @notice Send and announce an ERC20 payment to a stealth address
+   * @param _receiver Stealth address receiving the payment
+   * @param _tokenAddr Address of the ERC20 token being sent
+   * @param _amount Amount of the token to send, in its own base units
+   * @param _pkx X-coordinate of the ephemeral public key used to encrypt the payload
+   * @param _ciphertext Encrypted entropy (used to generated the stealth address) and payload extension
+   */
   function sendToken(
     address _receiver,
     address _tokenAddr,
     uint256 _amount,
     bytes32 _pkx, // ephemeral public key x coordinate
     bytes32 _ciphertext
-  ) public payable {
+  ) external payable {
     require(msg.value == toll, "Umbra: Must pay the exact toll");
-    require(tokenPayments[_receiver].amount == 0, "Umbra: Cannot send more tokens to stealth address");
+    require(tokenPayments[_receiver][_tokenAddr] == 0, "Umbra: Cannot send more tokens to stealth address");
 
-    tokenPayments[_receiver] = TokenPayment({token: _tokenAddr, amount: _amount});
+    tokenPayments[_receiver][_tokenAddr] = _amount;
     emit Announcement(_receiver, _amount, _tokenAddr, _pkx, _ciphertext);
 
     SafeERC20.safeTransferFrom(IERC20(_tokenAddr), msg.sender, address(this), _amount);
   }
 
-  function withdrawToken(address _acceptor) public {
-    uint256 amount = tokenPayments[msg.sender].amount;
-    address tokenAddr = tokenPayments[msg.sender].token;
+  // ======================================= Withdraw =============================================
+
+  /**
+   * @notice Withdraw an ERC20 token payment sent to a stealth address, which must be the sender
+   * @param _acceptor Address where withdrawn funds should be sent
+   * @param _tokenAddr Address of the ERC20 token being withdrawn
+   */
+  function withdrawToken(address _acceptor, address _tokenAddr) external {
+    uint256 amount = tokenPayments[msg.sender][_tokenAddr];
 
     require(amount > 0, "Umbra: No tokens available for withdrawal");
 
-    delete tokenPayments[msg.sender];
-    emit TokenWithdrawal(msg.sender, _acceptor, amount, tokenAddr);
+    delete tokenPayments[msg.sender][_tokenAddr];
+    emit TokenWithdrawal(msg.sender, _acceptor, amount, _tokenAddr);
 
-    SafeERC20.safeTransfer(IERC20(tokenAddr), _acceptor, amount);
+    SafeERC20.safeTransfer(IERC20(_tokenAddr), _acceptor, amount);
   }
 
+  /**
+   * @notice Withdraw an ERC20 token payment on behalf of a stealth address via signed authorization
+   * @param _stealthAddr The stealth address whose token balance will be withdrawn
+   * @param _acceptor Address where withdrawn funds should be sent
+   * @param _tokenAddr Address of the ERC20 token being withdrawn
+   * @param _sponsor Address which is compensated for submitting the withdrawal tx
+   * @param _sponsorFee Amount of the token to pay to the transfer
+   * @param _v ECSDA signature prefix
+   * @param _r ECSDA signature x-coordinate
+   * @param _s ECSDA signature y-coordinate
```

Comment:
These definitions aren't quite correct (seems you were thinking of uncompressed public keys), we can either use the ethers phrasing from [here](https://docs.ethers.io/v5/single-page/#/v5/api/utils/bytes/-%23-Signature), so:

```
   * @param _v ECSDA signature component: Parity of the `y` coordinate of point `R`
   * @param _r ECSDA signature component: x-coordinate of `R`
   * @param _s ECSDA signature component: `s` value of the signature
```

Or alternatively just lump everything into a `@dev` comment


```
   * @notice Withdraw an ERC20 token payment on behalf of a stealth address via signed authorization
   * @dev _v, _r, _s are from a valid secp256k1 signature from owner
   * @param _stealthAddr The stealth address whose token balance will be withdrawn
   * @param _acceptor Address where withdrawn funds should be sent
   * @param _tokenAddr Address of the ERC20 token being withdrawn
   * @param _sponsor Address which is compensated for submitting the withdrawal tx
   * @param _sponsorFee Amount of the token to pay to the transfer
   * @param _v ECSDA signature v-component
   * @param _r ECSDA signature r-component
   * @param _s ECSDA signature s-component

Reply:
Nice, thanks for the heads up here. I like the ethers phrasing 👍